extensions/amp-form/0.1/form-proxy.js

```diff
@@ -0,0 +1,361 @@
+ * @return {function(new:Object, !HTMLFormElement)}
+ */
+function getFormProxyConstr(win) {
+  if (!win.FormProxy) {
+    win.FormProxy = createFormProxyConstr(win);
+  }
+  return win.FormProxy;
+}
+
+
+/**
+ * @param {!Window} win
+ * @return {function(new:Object, !HTMLFormElement)}
+ */
+function createFormProxyConstr(win) {
+
+  /**
+   * @param {!HTMLFormElement} form
+   * @constructor
+   */
+  function FormProxy(form) {
+    /** @private @const {!HTMLFormElement} */
+    this.form_ = form;
+  }
+
+  const FormProxyProto = FormProxy.prototype;
+
+  // Hierarchy:
+  //   Node  <==  Element <== HTMLElement <== HTMLFormElement
+  //   EventTarget  <==  HTMLFormElement
+  const prototypes = [
+    win.HTMLFormElement.prototype,
+    win.HTMLElement.prototype,
+    win.Element.prototype,
+    win.Node.prototype,
+    win.EventTarget.prototype,
+  ];
+  prototypes.forEach(function(prototype) {
+    for (const name in prototype) {
+      const property = win.Object.getOwnPropertyDescriptor(prototype, name);
+      if (!property ||
+          // Exclude constants.
+          name.toUpperCase() == name ||
+          // Exclude on-events.
+          name.startsWith('on') ||
+          // Exclude properties that already been created.
+          win.Object.prototype.hasOwnProperty.call(FormProxyProto, name) ||
+          // Exclude some properties. Currently only used for testing.
+          blacklistedProperties && blacklistedProperties.indexOf(name) != -1) {
+        continue;
+      }
+      if (typeof property.value == 'function') {
+        // A method call. Call the original prototype method via `call`.
+        const method = property.value;
+        FormProxyProto[name] = function() {
+          return method.apply(this.form_, arguments);
+        };
+      } else {
+        // A read/write property. Call the original prototype getter/setter.
+        const spec = {};
+        if (property.get) {
+          spec.get = function() {
+            return property.get.call(this.form_);
+          };
+        }
+        if (property.set) {
+          spec.set = function(value) {
+            return property.set.call(this.form_, value);
+          };
+        }
+        win.Object.defineProperty(FormProxyProto, name, spec);
+      }
+    }
+  });
+
+  return FormProxy;
+}
+
+
+/**
+ * This is a very heavy-handed way to support browsers that do not have
+ * properties defined in the prototype chain. Specifically, this is necessary
+ * for Chrome 45 and under.
+ *
+ * See https://developers.google.com/web/updates/2015/04/DOM-attributes-now-on-the-prototype-chain
+ * for more info.
+ *
+ * @param {!HTMLFormElement} form
+ * @param {!Object} proxy
+ */
+function setupLegacyProxy(form, proxy) {
+  const proto = form.cloneNode(/* deep */ false);
+  for (const name in proto) {
+    if (name in proxy ||
+        // Exclude constants.
+        name.toUpperCase() == name ||
+        // Exclude on-events.
+        name.startsWith('on')) {
```

Comment:
You have to import our `startsWith` helper, `#startsWith` is only in recent ES6 browsers.